extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/KafkaDataSourceMetadata.java

```diff
@@ -21,133 +21,28 @@
-  }
-
-  @Override
-  public boolean matches(DataSourceMetadata other)
-  {
-    if (getClass() != other.getClass()) {
-      return false;
-    }
-
-    return plus(other).equals(other.plus(this));
-  }
-
-  @Override
-  public DataSourceMetadata plus(DataSourceMetadata other)
-  {
-    if (!(other instanceof KafkaDataSourceMetadata)) {
-      throw new IAE(
-          "Expected instance of %s, got %s",
-          KafkaDataSourceMetadata.class.getCanonicalName(),
-          other.getClass().getCanonicalName()
-      );
-    }
-
-    final KafkaDataSourceMetadata that = (KafkaDataSourceMetadata) other;
-
-    if (that.getKafkaPartitions().getTopic().equals(kafkaPartitions.getTopic())) {
-      // Same topic, merge offsets.
-      final Map<Integer, Long> newMap = new HashMap<>();
-
-      for (Map.Entry<Integer, Long> entry : kafkaPartitions.getPartitionOffsetMap().entrySet()) {
-        newMap.put(entry.getKey(), entry.getValue());
-      }
-
-      for (Map.Entry<Integer, Long> entry : that.getKafkaPartitions().getPartitionOffsetMap().entrySet()) {
-        newMap.put(entry.getKey(), entry.getValue());
-      }
-
-      return new KafkaDataSourceMetadata(new KafkaPartitions(kafkaPartitions.getTopic(), newMap));
-    } else {
-      // Different topic, prefer "other".
-      return other;
-    }
-  }
-
-  @Override
-  public DataSourceMetadata minus(DataSourceMetadata other)
-  {
-    if (!(other instanceof KafkaDataSourceMetadata)) {
-      throw new IAE(
-          "Expected instance of %s, got %s",
-          KafkaDataSourceMetadata.class.getCanonicalName(),
-          other.getClass().getCanonicalName()
-      );
-    }
-
-    final KafkaDataSourceMetadata that = (KafkaDataSourceMetadata) other;
-
-    if (that.getKafkaPartitions().getTopic().equals(kafkaPartitions.getTopic())) {
-      // Same topic, remove partitions present in "that" from "this"
-      final Map<Integer, Long> newMap = new HashMap<>();
-
-      for (Map.Entry<Integer, Long> entry : kafkaPartitions.getPartitionOffsetMap().entrySet()) {
-        if (!that.getKafkaPartitions().getPartitionOffsetMap().containsKey(entry.getKey())) {
-          newMap.put(entry.getKey(), entry.getValue());
-        }
-      }
-
-      return new KafkaDataSourceMetadata(new KafkaPartitions(kafkaPartitions.getTopic(), newMap));
-    } else {
-      // Different topic, prefer "this".
-      return this;
-    }
+    super(kafkaPartitions);
   }
 
   @Override
-  public boolean equals(Object o)
-  {
-    if (this == o) {
-      return true;
-    }
-    if (o == null || getClass() != o.getClass()) {
-      return false;
-    }
-    KafkaDataSourceMetadata that = (KafkaDataSourceMetadata) o;
-    return Objects.equals(kafkaPartitions, that.kafkaPartitions);
-  }
-
-  @Override
-  public int hashCode()
-  {
-    return Objects.hash(kafkaPartitions);
-  }
-
-  @Override
-  public String toString()
+  protected SeekableStreamDataSourceMetadata<Integer, Long> createConcreteDataSourceMetaData(
```

Comment:
what does concrete adds to the meaning here ? is there a non concrete DatasourceMetadata?